blog/_posts/2025-04-04-Traffic-Routing-Demystified-Best-practices-for-AKS-Load-Balancers.md

```diff
@@ -0,0 +1,141 @@
+
+![Preventing New Requests to Unhealthy Pods](./AKS/blog/assets/images/preventingnewrequeststoanunhealthypod.png)
+
+- **Immediate Health Check Response**:
+    As soon as a pod is marked for deletion, its `healthCheckNodePort` should start returning a 500 error. This signals to external load balancers that the pod is no longer healthy and should not receive new traffic.
+
+- **Load Balancer Probe Delay**:
+    External load balancers may take a few seconds to detect the unhealthy status of a pod. During this time, the pod might still receive traffic.
+
+To ensure your pods follow a similar timeline to gracefully shutdown, make sure to assess the following:
+
+1. **Delay Termination**:
+     After receiving the `TERM` signal, the application should wait for at least 10 seconds before proceeding with shutdown tasks (helps address the Load balancer probe delay). This can be achieved using a `preStop` hook in Kubernetes.
+
+    > **Note**: When using annotations `service.beta.kubernetes.io/azure-load-balancer-health-probe-interval` and/or `service.beta.kubernetes.io/azure-load-balancer-health-probe-num-of-probe`, consider changing the wait time to cover the annotation's needs (see [documentation](https://cloud-provider-azure.sigs.k8s.io/topics/loadbalancer/))
+
+2. **Announce Readiness as `false`**:
+     The application should update its readiness probe to indicate it is no longer ready to serve traffic. This allows Kubernetes to stop routing new requests to the pod.
+
+3. **Gracefully Close Existing Connections**:
+     The application should close all active connections, ensuring that no in-flight requests are dropped.
+
+4. **Exit the Process**:
+     Once all shutdown tasks are complete, the application should terminate its process cleanly.
+
+By implementing these best practices, you can minimize disruptions during pod shutdowns, maintain a seamless user experience, and ensure the reliability of your services in production environments.
+
+## Best Practices for Rolling Updates and Pod Rotation
+![Pod Rotation Timeline](./AKS/blog/assets/images/podrotationbestpractices.png)
+
+While the above work when a pod is being taken down in isolation, it does not cover cases like upgrades and rolling restarts which require coordination between the time the pod goes down and a new one comes up, ready to serve traffic. To optimize pod rotation, add the following best practice to your deployment:
+
+- **Set `minReadySeconds`**:
+    Configure the `minReadySeconds` parameter in your deployment to introduce a delay before Kubernetes is able to mark the pod as "available". This buffer gives the load balancer enough time to register the new pod and start routing traffic to it, while also preventing Kubernetes from deleting the old pod prematurely.
```

Comment:
Maybe worth describing the difference between deployment readiness and availability.  Readiness == the application says it's ready to receive new connections.  Availability == the application has been ready for long enough that a deployment rolling upgrade may continue to the next pod.

Reply:
Also, say what to set minReadySeconds to -- this is at least the time it takes for the external LB to probe the health endpoint and reconfigure its health.  I can't remember in AKS whether that's 5s or 10s by default (check with Mitch)